### PR TITLE
scientific notations added

### PR DESCRIPTION
**Pull request by yumuzi234**

### CODE REVIEW

File: pl/bare_func_test.go [hunk 1]
```diff
@@ -247,10 +248,13 @@ func TestBareFunc_bad(t *testing.T) {
 	// redefine
 	o("declConflict.var", "a:=3; a:=4")
 
-	o("argsMismatch", "printInt(true)")          // type mismatch
-	o("argsMismatch", "printInt(3, 4)")          // arg count mismatch
-	o("argsMismatch", "printInt()")              // arg count mismatch
-	o("argsMismatch", "a := printInt(3, 4)")     // mismatch
+	//argsMismatch
```
Comment (h8liu) [inline pl/bare_func_test.go]: please add space.

File: pl/bare_func_test.go [hunk 2]
```diff
@@ -284,7 +287,16 @@ func TestBareFunc_bad(t *testing.T) {
 	o("divideByZero", "a:=3/0")
 	o("divideByZero", "a:=3%0")
 
-	o("argsMismatch", "const a = -1; printUint(a)")
+	// build op
+	o("cannotCast.integerOverFlowed", "a:=12345678987654321")
+	o("cannotParseConst", "a:=19223372036854775808") // too large for int64
```
Comment (h8liu) [inline pl/bare_func_test.go]: this actually should be also overflow...

File: pl/parse/lex_ident.go [hunk 3]
```diff
@@ -10,25 +10,43 @@ func isLetter(r rune) bool {
...
+				x.Next()
+			}
+		}
+		if x.Rune() == 'e' || x.Rune() == 'E' {
+			isFloat = true
+			x.Next()
+			if lexing.IsDigit(x.Rune()) || x.Rune() == '-' {
+				x.Next()
+			}
+			for lexing.IsDigit(x.Rune()) {
```
Comment (h8liu) [inline pl/parse/lex_ident.go]: this will accept `3e-`, which is incorrect. please add a test case for this.

File: pl/sempass/operand.go [hunk 4]
```diff
@@ -14,15 +14,77 @@ import (
 func buildInt(b *builder, op *lexing.Token) tast.Expr {
 	ret, e := strconv.ParseInt(op.Lit, 0, 64)
 	if e != nil {
-		b.Errorf(op.Pos, "invalid integer: %s", e)
+		b.CodeErrorf(op.Pos, "pl.cannotParseConst",
```
Comment (h8liu) [inline pl/sempass/operand.go]: this should also be overflow.

File: pl/sempass/operand.go [hunk 5]
```diff
@@ -14,15 +14,77 @@ import (
...
+			"invalid integer: %s", e)
+		return nil
+	}
+	// ret can be a negative number? no?
```
Comment (h8liu) [inline pl/sempass/operand.go]: remove this comment.

File: pl/sempass/operand.go [hunk 6]
```diff
@@ -14,15 +14,77 @@ import (
...
+		b.CodeErrorf(op.Pos, "pl.cannotCast.integerOverFlowed",
+			"integer too large to fit in 32-bit")
+		return nil
+	}
+
+	ref := tast.NewConstRef(types.NewNumber(ret), ret)
+	return tast.NewConst(ref)
+}
+
+func buildFloat(b *builder, op *lexing.Token) tast.Expr {
```
Comment (h8liu) [inline pl/sempass/operand.go]: I would use v := strconv.ParseFloat(lit) and then first check if math.Floor(v) == v, and then check if uint32(v) is in range.

File: pl/bare_func_test.go (hunk 2 above)
Comment (yumuzi234) [inline pl/bare_func_test.go]: the error is returned by the strconv.ParseInt. overflow is only one case, maybe the only one will appear here?

File: pl/sempass/operand.go (hunk 6 above)
Comment (yumuzi234) [inline pl/sempass/operand.go]: OK

File: pl/bare_func_test.go (hunk 2 above)
Comment (h8liu) [inline pl/bare_func_test.go]: or use something like cannotCast.invalidFloat

my concern was that it is not in the "parsing" pass in the stage of the compiler.

File: pl/parse/lex_ident.go (hunk 3 above)
Comment (yumuzi234) [inline pl/parse/lex_ident.go]: it accept, like it will accept 1e, but there will be an error in buildFloat

Comment (h8liu) [inline pl/parse/lex_ident.go]: ok. please add test case to make sure that it will fail somewhere.

**Comment by yumuzi234** [comment]
added

**Comment by h8liu** [comment]
merged

**Pull request merged**